### PR TITLE
refactor and add LP copy editing to Weekly and Paper

### PR DESCRIPTION
**Pull request by johnduffell**
## What are you doing in this PR?

This adds the default promo code copy reading for the landing pages for paper and weekly.  Similar to the DS version here: https://github.com/guardian/support-frontend/pull/2955

It will look for `PAPER_LANDING_PAGE` and `WEEKLY_LANDING_PAGE` codes

## Why are you doing this?

Because marketing wanted it and it's 10% time so I decided to do it.

### CODE REVIEW

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 1]
```diff
@@ -3,174 +3,84 @@ package controllers
...
-    val hrefLangLinks = Map(
-      "en-us" -> buildCanonicalDigitalSubscriptionLink("us", orderIsAGift),
-      "en-gb" -> buildCanonicalDigitalSubscriptionLink("uk", orderIsAGift),
-      "en-au" -> buildCanonicalDigitalSubscriptionLink("au", orderIsAGift),
-      "en" -> buildCanonicalDigitalSubscriptionLink("int", orderIsAGift)
-    )
-    val shareImageUrl = Some("https://i.guim.co.uk/img/media/74422ad120c709448f433c34f5190e2465ffa65e/0_0_1200_1200/1200.png?width=1200&auto=format&fit=crop&quality=85&s=1407add4d016d15cc074b0f9de8f1433") // scalastyle:ignore
-    if (settings.switches.enableDigitalSubGifting.isOn || !orderIsAGift) {
+
+    if (!settings.switches.enableDigitalSubGifting.isOn && orderIsAGift) {
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:42]: this if statement logic was a bit upside down, so I flipped it for clarity.
i.e. `!(a || b) == (!a && !b)`
I guess we will get rid of the switch soon anyway?

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 2]
```diff
@@ -3,174 +3,84 @@ package controllers
...
+      Redirect(routes.DigitalSubscriptionController.digitalGeoRedirect(false)).withSettingsSurrogateKey
+    } else {
+      val canonicalLink = Some(buildCanonicalDigitalSubscriptionLink("uk", orderIsAGift))
+      val queryPromos = request.queryString.get("promoCode").map(_.toList).getOrElse(Nil)
       Ok(views.html.main(
-        title = title,
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:73]: most of these vals were only used once, and `title = title` didn't seem that helpful, so I condensed them down where they were relatively trivial

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 3]
```diff
@@ -3,174 +3,84 @@ package controllers
...
-        shareImageUrl = shareImageUrl,
+        hrefLangLinks = getPaperHrefLangLinks(orderIsAGift),
+        shareImageUrl = Some(
+          "https://i.guim.co.uk/img/media/74422ad120c709448f433c34f5190e2465ffa65e/0_0_1200_1200/1200.png" +
+            "?width=1200&auto=format&fit=crop&quality=85&s=1407add4d016d15cc074b0f9de8f1433"
+        ),
         shareUrl = canonicalLink
       ) {
-        Html(s"""<script type="text/javascript">
+        val maybePromotionCopy = landingCopyProvider.promotionCopy(queryPromos, DigitalPack, countryCode, DefaultPromotions.DigitalSubscription.landing)
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:62]: the `landingCopyProvider` is now shared between weekly/DS/paper

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 4]
```diff
@@ -3,174 +3,84 @@ package controllers
...
   landingCopyProvider: LandingCopyProvider,
   val assets: AssetsResolver,
   val actionRefiners: CustomActionBuilders,
-  identityService: IdentityService,
-  testUsers: TestUserService,
-  membersDataService: MembersDataService,
-  stripeConfigProvider: StripeConfigProvider,
-  payPalConfigProvider: PayPalConfigProvider,
   components: ControllerComponents,
   stringsConfig: StringsConfig,
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:25]: I'm not quite sure why we have StringsConfig, it just seems to indirectly read from the file referenced below in github?  Unless this was the precursor to the equivalent of the landingCopyProvider.
https://github.com/guardian/support-frontend/blob/6cebfa09877911211ad7e266d8dded59bd018562/support-frontend/conf/strings.conf#L1-L6

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 5]
```diff
@@ -3,174 +3,84 @@ package controllers
...
-    } else {
-      Redirect(routes.DigitalSubscriptionController.digitalGeoRedirect(false)).withSettingsSurrogateKey
-    }
-  }
-
-  def digitalGeoRedirect(orderIsAGift: Boolean = false): Action[AnyContent] = geoRedirect(
-    if (orderIsAGift) "subscribe/digital/gift" else "subscribe/digital"
-  )
-
-  def displayForm(orderIsAGift: Boolean): Action[AnyContent] = {
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:99]: I've moved the displayForm into a separate controller becaase this Controller had a huge number of parmeters and there didn't seem to be much connection between the two sides other than they happen to be the same product flow.
I did the same for weekly and paper, for consistency.

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 6]
```diff
@@ -3,174 +3,84 @@ package controllers
...
-      idUser,
-      uatMode,
-      priceSummaryServiceProvider.forUser(uatMode).getPrices(DigitalPack, promoCodes, readerType),
-      stripeConfigProvider.get(),
-      stripeConfigProvider.get(true),
-      payPalConfigProvider.get(),
-      payPalConfigProvider.get(true),
-      v2recaptchaConfigPublicKey,
-      orderIsAGift
+  private def getPaperHrefLangLinks(orderIsAGift: Boolean): Map[String, String] = {
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:75]: I extracted this function to align with weekly, as that's how it was done there and it makes sense.

File: support-frontend/app/controllers/DigitalSubscriptionController.scala [hunk 7]
```diff
@@ -179,19 +89,4 @@ class DigitalSubscriptionController(
 
 }
 
-class LandingCopyProvider(
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionController.scala:182]: moved to a new file (and extended to support weekly/paper)

File: support-frontend/app/controllers/DigitalSubscriptionFormController.scala [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  identityService: IdentityService,
+  testUsers: TestUserService,
+  membersDataService: MembersDataService,
+  stripeConfigProvider: StripeConfigProvider,
+  payPalConfigProvider: PayPalConfigProvider,
+  components: ControllerComponents,
+  settingsProvider: AllSettingsProvider,
+  fontLoaderBundle: Either[RefPath, StyleContent],
+  recaptchaConfigProvider: RecaptchaConfigProvider
+)(implicit val ec: ExecutionContext) extends AbstractController(components) with SettingsSurrogateKeySyntax {
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionFormController.scala:38]: there is still a lot of similarity between the three "formcontroller" but I haven't really touched in this PR.  Also, the main.scala.html used everywhere seems to have some required contributions specific stuff (e.g. amounts settings), but missing subs stuff (e.g. product prices and promo)  But that's not something for this PR.

File: support-frontend/app/controllers/DigitalSubscriptionFormController.scala [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+          },
+          user =>
+            if (orderIsAGift)
+              Future.successful(Ok(digitalSubscriptionFormHtml(user, orderIsAGift = true)))
+            else for (alreadyADigitalSubscriber <- userHasDigitalSubscription(membersDataService, request.user)) yield
+              if (alreadyADigitalSubscriber)
+                redirectToExistingThankYouPage
+              else
+                Ok(digitalSubscriptionFormHtml(user, orderIsAGift = false))
+        ).flatten.map(_.withSettingsSurrogateKey)
```
Comment (johnduffell) [inline support-frontend/app/controllers/DigitalSubscriptionFormController.scala:62]: the only change I made here is to flip so we only check if someone is an existing subscriber when necessary.  Before, for gift, we used to check, then ignore the result.

File: support-frontend/app/controllers/LandingCopyProvider.scala [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import com.gu.support.promotions.{ProductPromotionCopy, PromoCode, PromotionCopy, PromotionServiceProvider}
+
+class LandingCopyProvider(
+  promotionServiceProvider: PromotionServiceProvider,
+  stage: Stage
+) {
+
+  // To see if there is any promotional copy in place for this page we need to get a country in the current region (country group)
+  // this is because promotions apply to countries not regions. We can use any country however because the promo tool UI only deals
+  // with regions and then adds all the countries for that region to the promotion
```
Comment (johnduffell) [inline support-frontend/app/controllers/LandingCopyProvider.scala:16]: this comment was in the "paper" version of the code, and looked useful, so I retained it.

File: support-frontend/app/controllers/LandingCopyProvider.scala [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+      country <- countryGroup.countries.headOption
+    } yield country).getOrElse(UK)
+    promotionCopyForPrimaryCountry(queryPromos, product1, country, defaultCopyPromoCode)
+  }
+
+  def promotionCopyForPrimaryCountry(queryPromos: List[String], product1: Product, country: Country, defaultCopyPromoCode: PromoCode): Option[PromotionCopy] = {
+    val promoCode = queryPromos.headOption.getOrElse(defaultCopyPromoCode)
+    ProductPromotionCopy(promotionServiceProvider.forUser(false), stage)
+      .getCopyForPromoCode(promoCode, product1, country)
+  }
```
Comment (johnduffell) [inline support-frontend/app/controllers/LandingCopyProvider.scala:29]: this is called directly with `UK` by the paper one, as that's the only option

File: support-frontend/app/controllers/PaperSubscription.scala [hunk 12]
```diff
@@ -56,86 +35,37 @@ class PaperSubscription(
     Redirect(buildCanonicalPaperSubscriptionLink(withDelivery), request.queryString, status = FOUND)
   }
 
-  def paper(withDelivery: Boolean = false): Action[AnyContent] = CachedAction() { implicit request =>
```
Comment (johnduffell) [inline support-frontend/app/controllers/PaperSubscription.scala:59]: `withDelivery` wasn't actually used anywhere

File: support-frontend/app/controllers/DigitalSubscriptionController.scala (hunk 4 above)
Comment (rupertbates) [reply support-frontend/app/controllers/DigitalSubscriptionController.scala:25]: Hmm, no idea. Perhaps we can remove it

File: support-frontend/app/controllers/DigitalSubscriptionController.scala (hunk 5 above)
Comment (rupertbates) [reply support-frontend/app/controllers/DigitalSubscriptionController.scala:99]: 👏

**Review by rupertbates — APPROVED** [review]
Looks good, nice work moving the subs form stuff into a separate controller - that is overdue.

File: support-frontend/app/controllers/DigitalSubscriptionController.scala (hunk 4 above)
Comment (johnduffell) [reply support-frontend/app/controllers/DigitalSubscriptionController.scala:25]: let's do that "one day"

**Pull request merged**